### DISCUSSION TITLE
debug JSON not updating in PerspectiveWidget (and other discrepencies in perspective widget)

### DISCUSSION THREAD

**Discussion by jgunstone** [perspective-dev/perspective#2767] [Q&A]
### Steps to Reproduce:

```py
from datetime import date, datetime
import numpy as np
import pandas as pd
import perspective
from perspective import PerspectiveWidget, Aggregate, Sort, Plugin
import ipywidgets as w

data = pd.DataFrame({
    "int": np.arange(100),
    "float": [i * 1.5 for i in range(100)],
    "bool": [True for i in range(100)],
    "date": [date.today() for i in range(100)],
    "datetime": [datetime.now() for i in range(100)],
    "string": [str(i) for i in range(100)]
})

dg = perspective.PerspectiveWidget(
    data,
    plugin=Plugin.XBAR,
    aggregates={"datetime": Aggregate.ANY},
    sort=[["date", Sort.DESC]]
)
dg
```

 1. drag `bool` into "Group By" field
 2. no change in the debug JSON

### Expected Result:

when you make changes to the UI config settings they are updated in the debug JSON

### Actual Result:

![image](https://github.com/user-attachments/assets/6113780f-b440-44f5-92a5-6f9d0670bb4f)

### Environment:
fresh install - 

`environment.yml`
```yml
name: perspective-test
channels:
  - defaults
  - conda-forge
dependencies:
  - python
  - jupyterlab
  - perspective
```
```console
(base) jovyan@PORT4028:~/meetings-and-notes/20240925-PerspectiveGridTests$ mamba env create -f environment.yml
(base) jovyan@PORT4028:~/meetings-and-notes/20240925-PerspectiveGridTests$ mamba activate perspective-test
(perspective-test) jovyan@PORT4028:~/meetings-and-notes/20240925-PerspectiveGridTests$ mamba list perspective
# packages in environment at /home/jovyan/miniforge3/envs/perspective-test:
#
# Name                    Version                   Build  Channel
perspective               2.10.1          py311hcf300bf_2    conda-forge
```

### Additional Context:

some other issues I've encountered with the jupyter perspective widget vs the online examples (e.g. https://perspective.finos.org/block/?example=editable). True in both jupyter notebooks and the `PerspectiveCSV` viewer. 

- *datagrid* operation descriptions not shown

![image](https://github.com/user-attachments/assets/cecc3b41-b52d-480a-984f-90eb13f3338c)
![image](https://github.com/user-attachments/assets/427e76c9-c071-4d61-831e-9fa12e6534f0)

- reset button doing not doing anything

![image](https://github.com/user-attachments/assets/6af62547-755b-4f18-842d-dc332699388d)

**Comment by dhirschfeld** [comment]
Does the sort actually work for you? It didn't do anything for me when I tested with the `PerspectiveWidget`
```python
import pandas as pd
import numpy as np
from datetime import date, datetime
today = date.today().toordinal()
data = pd.DataFrame({
    "int": np.arange(100),
    "float": [i * 1.5 for i in range(100)],
    "bool": [True for i in range(100)],
    "date": [date.fromordinal(today + i) for i in range(100)],
    "datetime": [datetime.now() for i in range(100)],
    "string": [str(i) for i in range(100)]
})
PerspectiveWidget(data, sort=[["date", Sort.DESC]])
```

**Comment by dhirschfeld** [comment]
If `PerspectiveWidget` is being given an overhaul it might make sense to leverage the `anywidget` framework.
* https://anywidget.dev/en/getting-started/#what-is-anywidget
* https://blog.jupyter.org/anywidget-jupyter-widgets-made-easy-164eb2eae102

**Comment by texodus** [comment]
Your installation looks broken (at least - the theme CSS + web component + python library are from conflicting versions). I see (at least) two issues:

1) Conda is not a currently supported platform. The builds deployed there may work or not - I have not verified them. We are looking into adding support for this in the future but we won't be backdating support to old Perspective versions, speaking of which ..
2) Perspective 2.10.1 does not support JupyterLab 4. Perspective 3.1.0 is the latest version and does not behave as described.

(Converting this issue to a discussion)

**Comment by dhirschfeld** [comment]
As a data point, I just upgraded to 3.1 and sorting, reset and the debug panel seem to work as expected.

The correct syntax is now:
```python
PerspectiveWidget(data, sort=[["date", "desc"]])
```

**Comment by jgunstone** [comment]
hi - sorry - should've checked I was on the most recent version... 
though I would have expected this from a fresh install from: 

`environment.yml`
```yaml
name: perspective-test
channels:
  - defaults
  - conda-forge
dependencies:
  - python
  - jupyterlab
  - perspective
```

struggled to install v3.1.0
![image](https://github.com/user-attachments/assets/7898847c-27fa-4f94-86bd-5fca564b87a6)

also looks like only 3.0.1 available on conda
https://prefix.dev/channels/conda-forge/packages/perspective

just deleted the env and tried again with: 
`environment.yml`
```yaml
name: perspective-test
channels:
  - defaults
  - conda-forge
dependencies:
  - python
  - jupyterlab>4
  - perspective>3
```

and get the following error saying I required pyarrow. 
`mamba install pyarrow`

and then it seems to be working correctly :) 

```py
from datetime import date, datetime
import numpy as np
import pandas as pd
import perspective
from perspective import PerspectiveWidget
import ipywidgets as w

data = pd.DataFrame({
    "int": np.arange(100),
    "float": [i * 1.5 for i in range(100)],
    "bool": [True for i in range(100)],
    "date": [date.today() for i in range(100)],
    "datetime": [datetime.now() for i in range(100)],
    "string": [str(i) for i in range(100)]
})
data.to_csv("data.csv")
dg = perspective.PerspectiveWidget(
    data,
)
dg
```

if I now try and open `data.csv` in the filebrowser it doesn't work:
![image](https://github.com/user-attachments/assets/5c2fad3f-8a69-49a2-befc-e76c20576fdc)

(maybe this is fixed in the most recent version?) 



pyarrow error:
![image](https://github.com/user-attachments/assets/8149848d-bd39-495b-be61-2de4ae9a65de)

**Comment by texodus** [comment]
> struggled to install v3.1.0

The name of the PYPI package is `perspective-python`, not `perspective`.

> also looks like only 3.0.1 available on conda

Again, ***_Perspective does not support Conda yet_.*** This includes mamba.
